### ISSUE TITLE
ICE when deriving `CheckBytes` over a `RefAsBox` slice field.

### ISSUE THREAD

**Issue by tobz** [rkyv/rkyv#221]
MCVE:

```rust
#[derive(Archive, Serialize, Debug)]
#[archive_attr(derive(CheckBytes))]
pub struct Record<'a> {
    /// The checksum of the record.
    ///
    /// The checksum is CRC32C(big_endian_bytes(id) + payload).
    checksum: u32,
    /// The record ID.
    ///
    /// This is monotonic across records.
    id: u64,
    // The record length.
    //
    // This is the number of bytes that follow the header.
    #[with(RefAsBox)]
    payload: &'a [u8],
}
```

ICE:

> error: internal compiler error: compiler/rustc_traits/src/normalize_erasing_regions.rs:54:32: could not fully normalize `std::result::Result<&<rkyv::with::With<&[u8], rkyv::with::RefAsBox> as rkyv::Archive>::Archived, <<rkyv::with::With<&[u8], rkyv::with::RefAsBox> as rkyv::Archive>::Archived as bytecheck::CheckBytes<__C>>::Error>

**Comment by djkoloski** [comment]
I'll minimize this and report it upstream.

**Comment by mamins1376** [comment]
I filed related two ICEs ([this](https://github.com/rust-lang/rust/issues/91231) and [this](https://github.com/rust-lang/rust/issues/91234)) which the latter causes this code not to compile:

```Rust
#[derive(Archive)]
enum Request<'a> {
    Variant(&'a usize)
}
```

**Comment by mamins1376** [comment]
Fixed in https://github.com/rust-lang/rust/pull/91255

**Comment by djkoloski** [comment]
I'll make sure that all of these are fixed with the next Rust release (1.59) and reopen if any of them return.
